### PR TITLE
Fix content disposition parsing

### LINKED ISSUE #75: Filenames parsed from the `Content-Disposition` headers are incorrect.
Hi there. We've recently identified an issue with this library when using the downloader alongside the JFrog Artifactory platform. This doesn't affect only this platform -- it could affect many platforms that use the aforementioned header.

If no `fileName` option is provided to **node-downloader-helper**, the helper will attempt to parse the file name from the `Content-Disposition` header. If the header is present, which it is for files downloaded from Artifactory, it can contain more than what the script expects.

This gives the file a mangled name. Incidentally, this can often pass on macOS/Unix, but will cause the download to fail on Windows.

As you can see in this function:

https://github.com/hgouveia/node-downloader-helper/blob/33aab440ce6d5bf996ffec07fbadad61862811c6/src/index.js#L658-L683

on this line:
https://github.com/hgouveia/node-downloader-helper/blob/33aab440ce6d5bf996ffec07fbadad61862811c6/src/index.js#L667

The entirety of the header is included in this substring. There can be multiple properties in this header, as per spec. I think this header parsing needs to be done a little differently, perhaps with some sort of token splitting.

I have an example script I've been toying with that might have a potential solution:
```javascript
const headers = {
  'content-disposition': `attachment; filename="Setup64.exe"; filename*=UTF-8''Setup64.exe`
};

const parseHeaderValue = (headerValue) => {
  let headerDictionary = headerValue.split(';');
  headerDictionary = headerDictionary.reduce((acc, val) => {
    const splitVal = val.trim().split('=');
    return { ...acc, [splitVal[0]]: splitVal[1] ?? '' };
  }, {});
  return headerDictionary;
}


const contentDispositionValues = parseHeaderValue(headers['content-disposition']);

if(headers.hasOwnProperty('content-disposition') &&
            contentDispositionValues['filename']) {
  let fileName = headers['content-disposition'];

  fileName = fileName.trim();

  fileName = contentDispositionValues['filename'];

  fileName = fileName.replace(new RegExp('"', 'g'), '');

  fileName = fileName.replace(/[/\\]/g, '');
  
  console.log(fileName);
}
```

In our codebase, we solved this problem by always passing in a `fileName` optional property when constructing the `DownloaderHelper`, but I'm hoping this will help eliminate a frustrating edge case we experienced. Thank you :)

### PR DESCRIPTION
**Pull request by Chaphasilor**
this closes #75

if you want me to change anything, just let me know!

### CODE REVIEW

**Comment by Chaphasilor** [comment]
hold on, this builds against ***Node 6***‽  
Maybe it's time to update the build target @hgouveia, the legacy `url` module is [discontinued](https://nodejs.org/docs/latest-v17.x/api/url.html#legacy-url-api). It even was deprecated for some time...  
I can fix the error for now, but we should really think about this :)

**Review by paul-jarrow — CHANGES_REQUESTED** [review]
It's a bit difficult to read the submission of `src/index.js`, it appears to be a change on the entire file.

**Comment by Chaphasilor** [comment]
You're absolutely right, I think what happened is that the line endings got converted...

**Comment by Chaphasilor** [comment]
@paul-jarrow better now? :D

File: src/index.js [hunk 1]
```diff
@@ -658,20 +658,37 @@ export class DownloaderHelper extends EventEmitter {
     __getFileNameFromHeaders(headers, response) {
         let fileName = '';
 
+        const fileNameAndEncodingRegExp = /.*filename\*=.*?''([^"].+?[^"])(?:(?:;)|$)/i // match everything after the specified encoding behind a case-insensitive `filename*=`
+        const fileNameWithQuotesRegExp = /.*filename="(.*?)";?/i // match everything inside the quotes behind a case-insensitive `filename=`
+        const fileNameWithoutQuotesRegExp = /.*filename=([^"].+?[^"])(?:(?:;)|$)/i // match everything immediately after `filename=` that isn't surrounded by quotes and is followed by either a `;` or the end of the string
+        
+        const ContentDispositionHeaderExists = headers.hasOwnProperty('content-disposition')
+        const fileNameAndEncodingMatch = !ContentDispositionHeaderExists ? null : headers['content-disposition'].match(fileNameAndEncodingRegExp)
```
Comment (paul-jarrow) [inline src/index.js:666]: One small change, if I'm understanding [this RFC](https://datatracker.ietf.org/doc/html/rfc5987#section-3.2.2) document correctly, the `filename*=` encoding property can have another element that isn't accounted for in the matcher, like so:  
`filename*=iso-8859-1'en'%A3%20rates`

The key part being the `en`. I think we need this matcher instead:
```regex
/.*filename\*=.*?'.*?'([^"].+?[^"])(?:(?:;)|$)/i
```

Which adds `.*?` in between the single quotes.

File: test/index.spec.js [hunk 2]
```diff
@@ -258,6 +280,40 @@ describe('DownloaderHelper', function () {
...
+                {header: `attachment; filename*= UTF-8''%e2%82%ac%20rates`, fileName: '%e2%82%ac%20rates'},
+                // eslint-disable-next-line quotes
+                {header: `attachment;filename*=utf-8''%e2%82%ac%20rates;filename="EURO rates"`, fileName: '%e2%82%ac%20rates'},
+                // eslint-disable-next-line quotes
+                {header: `attachment;filename*=utf-8''%e2%82%ac%20rates;filename=EURO rates`, fileName: '%e2%82%ac%20rates'},
+                // eslint-disable-next-line quotes
+                {header: `attachment;filename*=utf-8''%e2%82%ac% 20rates;filename=EURO rates   `, fileName: '%e2%82%ac% 20rates'},
+                // eslint-disable-next-line quotes
+                {header: `attachment;fIlEnAmE=EURO rates`, fileName: 'EURO rates'},
+            ]
```
Comment (paul-jarrow) [inline test/index.spec.js:309]: Pretty comprehensive! I'd just add a test for the new change:
```javascript
// eslint-disable-next-line quotes
{header: `attachment; filename*=UTF-8'en'Setup64.exe`, fileName: 'Setup64.exe'},
```

**Review by paul-jarrow — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 1 above)
Comment (Chaphasilor) [reply src/index.js:666]: Ah, I was wondering about those quotes, looked like a strange syntax to me!  
Good catch, I'll change it later :)

**Comment by hgouveia** [comment]
@Chaphasilor feel free to update the package to at least node 14 , and remove the legacyURL, node 6 was a requirement for an old project i was working on that uses this library and it wasnt possible to upgrade, but now we finally moved on to a newer backend

**Comment by Chaphasilor** [comment]
> @Chaphasilor feel free to update the package to at least node 14 , and remove the legacyURL

@hgouveia okay, I'll do that in a different PR and try to clean up some of the older and deprecated things. maybe we should release it as a new major version (2.0.0) to make users aware of the breaking changes, but we can talk about it some more once I get around to the PR :)

**Comment by hgouveia** [comment]
@Chaphasilor agree lets up the version as suggested

**Review by paul-jarrow — APPROVED** [review]

**Comment by Chaphasilor** [comment]
@paul-jarrow I think if you don't see any other problems, @hgouveia could merge this now. maybe try to install the package from my branch via npm (` npm i git+https://github.com/Chaphasilor/node-downloader-helper.git#fix-content-disposition-parsing`) and do some real-world testing, just to be sure 😁

**Comment by hgouveia** [comment]
I will do some checking tomorrow before merge since I am not in the pc at the moment, but if @paul-jarrow could do some checking as well, if all fine I might also publish the new version tomorrow too since I think with all the changes from @Chaphasilor already deserve a version hehe

**Comment by paul-jarrow** [comment]
> I will do some checking tomorrow before merge since I am not in the pc at the moment, but if @paul-jarrow could do some checking as well, if all fine I might also publish the new version tomorrow too since I think with all the changes from @Chaphasilor already deserve a version hehe

I think that I've done the checking I need to, thus the acceptance. It seems to work for the cases I can think of!

**Pull request merged**